vm/src/builtins/bytearray.rs

```diff
@@ -667,40 +663,42 @@ impl Comparable for PyByteArray {
 impl BufferProtocol for PyByteArray {
     fn get_buffer(zelf: &PyRef<Self>, _vm: &VirtualMachine) -> PyResult<Box<dyn Buffer>> {
         zelf.exports.fetch_add(1);
-        Ok(Box::new(zelf.clone()))
+        let buf = ByteArrayBuffer {
+            bytearray: zelf.clone(),
+            options: BufferOptions {
+                readonly: false,
+                len: zelf.len(),
+                ..Default::default()
+            },
+        };
+        Ok(Box::new(buf))
     }
 }
 
-impl Buffer for PyByteArrayRef {
+#[derive(Debug)]
+struct ByteArrayBuffer {
+    bytearray: PyByteArrayRef,
+    options: BufferOptions,
+}
+
+impl Buffer for ByteArrayBuffer {
     fn obj_bytes(&self) -> BorrowedValue<[u8]> {
-        PyRwLockReadGuard::map(self.borrow_value(), |x| x.elements.as_slice()).into()
+        PyRwLockReadGuard::map(self.bytearray.borrow_value(), |x| x.elements.as_slice()).into()
     }
 
     fn obj_bytes_mut(&self) -> BorrowedValueMut<[u8]> {
-        PyRwLockWriteGuard::map(self.borrow_value_mut(), |x| x.elements.as_mut_slice()).into()
+        PyRwLockWriteGuard::map(self.bytearray.borrow_value_mut(), |x| {
+            x.elements.as_mut_slice()
+        })
+        .into()
     }
 
     fn release(&self) {
-        let mut w = self.buffer_options.write();
-        if self.exports.fetch_sub(1) == 1 {
-            *w = None;
-        }
+        self.bytearray.exports.fetch_sub(1);
     }
 
     fn get_options(&self) -> BorrowedValue<BufferOptions> {
-        let guard = self.buffer_options.upgradable_read();
-        let guard = if guard.is_none() {
-            let mut w = PyRwLockUpgradableReadGuard::upgrade(guard);
-            *w = Some(Box::new(BufferOptions {
-                readonly: false,
-                len: self.len(),
-                ..Default::default()
-            }));
-            PyRwLockWriteGuard::downgrade(w)
-        } else {
-            PyRwLockUpgradableReadGuard::downgrade(guard)
-        };
-        PyRwLockReadGuard::map(guard, |x| x.as_ref().unwrap().as_ref()).into()
+        (&self.options).into()
     }
 }
```

Comment:
I am thinking maybe simpler way is change the trait fn get_options to return a reference instead the borrow, than we need create the Buffer struct with options for every classes.